### PR TITLE
[DNM] *: support the local transaction

### LINKED ISSUE #20853: Start global transactions for some special SQLs
## Development Task

Some special SQLs visit metadata or they are executed in background sessions, so they should always begin as a global transaction.
This is a part of the global/local transaction roadmap #20448.

## Rationale

The special SQLs mentioned above include:
- Statements that write metadata, e.g. DDL, select nextval(seq), insert with an auto-increment column. Obtaining the next value from an auto-increment column or sequence may trigger an extra transaction, which writes the metadata to update the next start value. These statements create new KV transactions to write metadata, which is global and cannot be bound to any DC, so we need to guarantee that the new transactions are global.
- Statements(except for DMLs) that write system tables, e.g. analyze tables, grant privileges, set global variables, create bindings. These statements create new transactions in background system sessions to write system tables. a.k.a. internal SQLs. We need to guarantee that the new transactions are global.
- DDLs that write user tables, e.g. create/drop indexes, add columns, recover/cleanup indexes. Creating indexes and adding columns do the backfill jobs in background worker goroutines. Recover indexes and cleanup indexes do the backfill jobs in a user session. They all create new KV transactions to write user tables. We need to guarantee that all the new transactions are global.
- DMLs that write system tables, e.g. update mysql.tidb directly. These statements won't create new transactions, but they will fail as described in #20827. It's acceptable.

To summarize, we need to consider 3 scenarios:
- Internal SQLs. They always run in system sessions.
- SQLs that write metadata. They may run in user sessions.
- DDLs that write user tables. They may run in user sessions.

## Implementations

Internal SQLs can be recognized by `SessionVars.InRestrictedSQL` or `session.isInternal()`. Internal statements are executed through `session.ExecuteInternal()` or `execRestrictedSQL()`.

Transactions that write metadata are executed through `Meta.txn` API. We need to guarantee that the `kv.Transaction` passed to `NewMeta` is started as a global transaction.

DDLs that write user tables can be found in `RecoverIndexExec`, `CleanupIndexExec`, `addIndexWorker`, `cleanUpIndexWorker`, `updateColumnWorker`, etc.

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

Part of #20448, close #20853.

### What is changed and how it works?

What's Changed:

* Prepare more than one `TSFuture` for both local and global transaction scopes.
* Start the transaction with `TxnScope` by default.
* Start global transactions for some special SQLs:
    - Check `session.isInternal()` for the internal SQLs.
    - Start the global transactions for the `Meta.txn`.
    - Set `txn_scope` to "global" before a DDL executor creates a transaction.

The main changes are all in the `session.go`. You can review this part first because the other changes bascially are changes to the interface parameters and some specific logic to the special SQLs such as DDL.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- No release note.

### CODE REVIEW

**Comment by JmPotato** [comment]
Some to-dos:

- [ ] Garbage collect the useless `TxnScope` for the `session.txns`.
- [ ] Refine the codes.
- [ ] Add some tests.
- [ ] Update the `check_dev_2` to support the Local TSO/Txn.

**Comment by JmPotato** [comment]
@djshow832 @bobotu PTAL. The main changes are focused on the `session/session.go`, I'm not quite sure if this is the best solution, though it works. If the idea is ok, I will further add some related tests then.

File: kv/txn.go [hunk 1]
```diff
@@ -26,14 +26,14 @@ import (
 )
 
 // RunInNewTxn will run the f in a new transaction environment.
-func RunInNewTxn(store Storage, retryable bool, f func(txn Transaction) error) error {
+func RunInNewTxn(store Storage, txnScope string, retryable bool, f func(txn Transaction) error) error {
```
Comment (disksing) [inline kv/txn.go:29]: Could we add another function to allow using local txnScope so we don't have to pass in `GlobalTxnScope` everywhere?

File: session/session.go [hunk 2]
```diff
@@ -160,7 +161,9 @@ func (h *StmtHistory) Count() int {
 type session struct {
 	// processInfo is used by ShowProcess(), and should be modified atomically.
 	processInfo atomic.Value
-	txn         TxnState
+	// txns holds multiple txns for different txn scopes, which will be like:
+	// txn_scope -> TxnState
+	txns map[string]*TxnState
```
Comment (djshow832) [inline session/session.go:166]: Is there more than one transaction in a session at the same time?

Comment (JmPotato) [reply session/session.go:166]: Yes, we will prepare at least two transactions in a session when we set the `txn_scope` to non-global, and usually one is global and another is local. But it's also possible to have more than two transactions if the user changes the session variable `txn_scope` to different values frequently. I'm planning to do some GC work to handle these useless transactions.

File: bindinfo/handle.go [hunk 3]
```diff
@@ -213,7 +213,9 @@ func (h *BindHandle) CreateBindRecord(sctx sessionctx.Context, record *BindRecor
 		h.bindInfo.Unlock()
 	}()
 
-	txn, err1 := h.sctx.Context.Txn(true)
+	txn, err1 := h.sctx.Context.Txn(true, func(txnOpt *sessionctx.TxnOption) {
```
Comment (bobotu) [inline bindinfo/handle.go:216]: I think we can define this closure as a global variable and use that variable. In most cases, we just need to enable `GlobalTxnScope`, the boilerplate is too annoying.

**Review by bobotu — CHANGES_REQUESTED** [review]

**Comment by JmPotato** [comment]
I will split this PR into two parts, and the first part of them is #21345. Please take a look at that PR. Thanks! @Yisaer @bobotu @nolouch @djshow832 @disksing

**Pull request closed without merging**